zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -611,48 +590,61 @@ private[consumer] final class Runloop(
       _ => command.succeed.as(state.copy(subscription = command.subscription))
     )
 
-  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Unit]] =
-    ZStream
-      .mergeAll(3, 1)(
-        ZStream.repeatWithSchedule(Command.Poll, Schedule.fixed(pollFrequency)),
-        ZStream.fromQueue(requestQueue).mapChunks(c => Chunk.single(Command.Requests(c))),
-        ZStream.fromQueue(commandQueue)
-      )
-      .tap(cmd => diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd)))
-      .runFoldZIO(State.initial) { (state, cmd) =>
-        ZIO.ifZIO(isShutdown)(onTrue = handleShutdown(state, cmd), onFalse = handleOperational(state, cmd))
-      }
+  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Any]] =
+    pollLoop(false, State.initial)
       .tapErrorCause(cause => ZIO.logErrorCause("Error in Runloop", cause))
       .onError(cause => partitions.offer(Take.failCause(cause)))
-      .unit
       .forkScoped
+
+  private def pollLoop(initialized: Boolean, state: State): ZIO[Any, Throwable, State] = {
+    def processCommands(state: State) = for {
+      // Gather available commands or return immediately if nothing in the queue
+      commands   <- commandQueue.takeAll
+      isShutdown <- isShutdown
+      updatedState <- ZIO.foldLeft(commands)(state) { case (s, cmd) =>
+                        (if (isShutdown) handleShutdown(s, cmd) else handleOperational(s, cmd)) <*
+                          diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd))
+                      }
```

Comment:
`isShutdown` is not going change while looping over the commands.

```suggestion
      isShutdown <- isShutdown
      updatedState <- if (isShutdown) {
        ZIO.foldLeft(commands)(state) { case (s, cmd) => handleShutdown(s, cmd)) <* diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd)) }
      } else {
        ZIO.foldLeft(commands)(state) { case (s, cmd) => handleOperational(s, cmd)) <* diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd)) }
      }
```

Can be further improved by extracting `diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd))` to a helper method:

Something like:

```suggestion
  def emitCmdDiagnostics(cmd: Command) = diagnostics.emitIfEnabled(DiagnosticEvent.RunloopEvent(cmd))

      isShutdown <- isShutdown
      updatedState <- if (isShutdown) {
        ZIO.foldLeft(commands)(state) { case (s, cmd) => handleShutdown(s, cmd)) <* emitCmdDiagnostics(cmd) }
      } else {
        ZIO.foldLeft(commands)(state) { case (s, cmd) => handleOperational(s, cmd)) <* emitCmdDiagnostics(cmd) }
      }
```

Reply:
Not sure if this makes things more readable..